tools/CheckDates.scala

```diff
@@ -0,0 +1,184 @@
+    If the message of last change match regex "\\[*.\\]" (e.g. '[Tech message]')  it will be ignored and 
+    we will take next commit.
+
+    analyzeDates - just shows if anything is ok
+    fixWrongDates - actual fixing
+
+    From root folder:
+    Usage: scala tools/CheckDates.scala analyzeDates
+           scala tools/CheckDates.scala fixWrongDates
+
+  """
+  trait DateComparison {
+    def actualLastUpdateYear: String
+    def gitLastUpdateYear: String
+    def actualCreationYear: String
+    def gitCreationYear: String
+
+    def ifUpdateDateWrong(): Boolean = {
+      actualLastUpdateYear.toInt != gitLastUpdateYear.toInt
+    }
+
+    def ifUpdateDateWrongStr(): Boolean = {
+      actualLastUpdateYear != gitLastUpdateYear
+    }
+
+    def ifOneYear(): Boolean = {
+      gitCreationYear != gitLastUpdateYear
+    }
+  }
+
+  case class FileDates(pathToFile: String,
+                      actualLastUpdateYear: String,
+                      gitLastUpdateYear: String,
+                      actualCreationYear: String,
+                      gitCreationYear: String) extends DateComparison
+
+  case class GitInfo(message: String, date: String)
+
+  def findRootFolder(): String = {
+    val wd = (s"pwd") !!
+    var root_folder = wd
+
+    if (wd.split("/").contains("junit")) {
+       root_folder =  wd.split("/").dropRight(4).mkString("/")
+     }
+    root_folder
+  }
+
+  def readFilesList(): Array[String] = {
+    var rootFolder = findRootFolder()
+    val myval =  (s"find ${rootFolder} -name *.scala ! -path ./tools/*  -type f") !!
+    val files_array = myval.split("\n").map(_.toString).distinct
+    files_array
+  }
+
+  def readFile(pathToFile: String): String = {
+    val source = scala.io.Source.fromFile(pathToFile)("UTF-8")
+    val lines = try source.mkString finally source.close()
+    lines
+  }
+
+  def ifFileWasInGit(pathToFile: String): Boolean = {
+    val res_dates = (s"git log $pathToFile" !!)
+    if(res_dates == "") {
+      false
+    } else {
+      true
+    }
+  }
+
+  def getGitCreationYear(pathToFile: String): String = {
+    val buf_dates = (s"git log   --format=%cd $pathToFile" !!)
+    val dates = buf_dates.split("\\n").distinct
+    val res = dates.last.split(" ")(4)
+    res
+  }
+
+  def getGitLastUpdateYear(pathToFile: String): String = {
+    val buf_dates = (s"git log -3  --format=%cd $pathToFile" !!)
+    val buf_messages = (s"git log -3  $pathToFile" !!)
+    val messages = buf_messages.split("commit").distinct.drop(1)
+    val dates = buf_dates.split("\\n").distinct
+    val gits_info = (messages, dates).zipped.map{ case(a,b) => GitInfo.tupled((a,b))}
+    val res = gits_info.filter(!_.message.contains("[Dates fixing]"))
+    res(0).date.split(" ")(4)
+  }
+
+  def getYear(pathToFile: String, lower_bound: Int, upper_bound: Int): String = {
+    val file = readFile(pathToFile)
+    val start_index = file.indexOf("(c) 20")
```

Comment:
Suggest applying a limit to how far into the file the substring can be search for. Since the copyright header has a known format I'd be inclined to look in line 3 only,

```
/*                     __                                               *\
**     ________ ___   / /  ___     Scala API                            **
**    / __/ __// _ | / /  / _ |    (c) 2003-2013, LAMP/EPFL             **
**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
** /____/\___/_/ |_/____/_/ | |                                         **
**                          |/                                          **
\*                                                                      */
```

You might be interesting in backlogging other types of copyright header or discussing a move to a uniform header.

```
/* NSC -- new Scala compiler
 * Copyright 2007-2013 LAMP/EPFL
 * @author  Manohar Jonnalagedda
 */
```